File: Source/Error/CNError.m
/**
 * @file	CNError.m
 * @brief	Extend NSError class
 * @par Copyright
 *   Copyright (C) 2014 Steel Wheels Project
 */

#import "CNError.h"

static NSString * s_errorDomain = @"github.com.steelwheels.Coconut" ;
static NSString * s_errorLocationKey = @"errorLocation" ;

@implementation NSError (CNErrorExtension)

+ (NSError *) parseErrorWithMessage: (NSString *) message
{
	NSDictionary * errorinfo = @{NSLocalizedDescriptionKey: message};
	return [[NSError alloc] initWithDomain: s_errorDomain
					  code: CNParserError
				      userInfo: errorinfo] ;
}

+ (NSError *) parseErrorWithMessage: (NSString *) message withLocation: (CNErrorLocation *) location
{
	NSDictionary * errorinfo = @{NSLocalizedDescriptionKey: message,
				     s_errorLocationKey: location} ;
	return [[NSError alloc] initWithDomain: s_errorDomain
					  code: CNParserError
				      userInfo: errorinfo] ;
}

+ (NSError *) fileErrorWithMessage: (NSString *) message
{
	NSDictionary * errorinfo = @{NSLocalizedDescriptionKey: message};
	return [[NSError alloc] initWithDomain: s_errorDomain
					  code: CNFileError
				      userInfo: errorinfo] ;
}

+ (NSError *) memoryErrorWithMessage: (NSString *) message atFunction: (const char *) funcname
{
	if(funcname){
		NSString * filepath = [[NSString alloc] initWithFormat: @" at function \"%s\"", funcname] ;
		message = [message stringByAppendingString: filepath] ;
	}
	NSDictionary * errorinfo = @{NSLocalizedDescriptionKey: message};
	return [[NSError alloc] initWithDomain: s_errorDomain
					  code: CNMemoryError
				      userInfo: errorinfo] ;
}

+ (NSError *) memoryErrorWithKind: (CNMemoryErrorKind) kind atFunction: (const char *) funcname
{
	NSString * message = nil ;
	switch(kind){
		case CNMemoryAllocationError: {
			message = @"Memory allocation error" ;
		} break ;
		default: {
			message = @"Unkown memory error" ;
		} break ;
	}
	return [NSError memoryErrorWithMessage: message atFunction: funcname] ;
}

- (void) printToFile: (FILE *) outfp
{
	fputs("[Error] ", outfp) ;
	NSDictionary * userinfo = [self userInfo] ;
	NSString * desc = [userinfo valueForKey: NSLocalizedDescriptionKey] ;
	if(desc != nil){
		fprintf(outfp, "%s", [desc UTF8String]) ;
	}
	fputc('\n', outfp) ;
}

@end

File: CoconutData/Source/Error/CNError.swift
/**
 * @file	CNErrorExtension.swift
 * @brief	Extend NSError class
 * @par Copyright
 *   Copyright (C) 2015 Steel Wheels Project
 */

import Foundation

public enum CNErrorCode: Int {
	case Information	= 0
	case InternalError	= 1
	case ParseError		= 2
	case FileError		= 3
	case SerializeError	= 4
	case UnknownError	= 5
}

public extension NSError
{
	class func domain() -> String {
		return "github.com.steelwheels.Coconut"
	}

	class func errorLocationKey() -> String {
		return "errorLocation"
	}

	class func informationNotice(message m: String) -> NSError {
		let userinfo = [NSLocalizedDescriptionKey: m]
		return NSError(domain: self.domain(), code: CNErrorCode.Information.rawValue, userInfo: userinfo)
	}

	class func internalError(message m: String) -> NSError {
		let userinfo = [NSLocalizedDescriptionKey: m]
		return NSError(domain: self.domain(), code: CNErrorCode.InternalError.rawValue, userInfo: userinfo)
	}

	class func internalError(message m: String, location l: NSString) -> NSError {
		let userinfo = [NSLocalizedDescriptionKey: NSString(string: m), self.errorLocationKey(): l]
		return NSError(domain: self.domain(), code: CNErrorCode.InternalError.rawValue, userInfo: userinfo)
	}

	class func parseError(message m: String) -> NSError {
		let userinfo = [NSLocalizedDescriptionKey: m]
		return NSError(domain: self.domain(), code: CNErrorCode.ParseError.rawValue, userInfo: userinfo)
	}

	class func parseError(message m: String, location l: NSString) -> NSError {
		let userinfo = [NSLocalizedDescriptionKey: NSString(string: m), self.errorLocationKey(): l]
		return NSError(domain: self.domain(), code: CNErrorCode.ParseError.rawValue, userInfo: userinfo)
	}

	class func fileError(message m: String) -> NSError {
		let userinfo = [NSLocalizedDescriptionKey: m]
		let error = NSError(domain: self.domain(), code: CNErrorCode.FileError.rawValue, userInfo: userinfo)
		return error
	}

	class func fileError(message m: String, location l: NSString) -> NSError {
		let userinfo = [NSLocalizedDescriptionKey: NSString(string: m), self.errorLocationKey(): l]
		return NSError(domain: self.domain(), code: CNErrorCode.FileError.rawValue, userInfo: userinfo)
	}

	class func serializeError(message m: String) -> NSError {
		let userinfo = [NSLocalizedDescriptionKey: m]
		return NSError(domain: self.domain(), code: CNErrorCode.SerializeError.rawValue, userInfo: userinfo)
	}

	class func serializeError(message m: String, location l: NSString) -> NSError {
		let userinfo = [NSLocalizedDescriptionKey: NSString(string: m), self.errorLocationKey(): l]
		return NSError(domain: self.domain(), code: CNErrorCode.SerializeError.rawValue, userInfo: userinfo)
	}

	class func unknownError() -> NSError {
		let userinfo = [NSLocalizedDescriptionKey: "Unknown error"]
		return NSError(domain: self.domain(), code: CNErrorCode.UnknownError.rawValue, userInfo: userinfo)
	}

	class func unknownError(location l: NSString) -> NSError {
		let userinfo = [NSLocalizedDescriptionKey: NSString(string: "Unknown error"), self.errorLocationKey(): l]
		return NSError(domain: self.domain(), code: CNErrorCode.UnknownError.rawValue, userInfo: userinfo)
	}

	var errorCode: CNErrorCode {
		get {
			if let ecode = CNErrorCode(rawValue: code) {
				return ecode
			} else {
				return .UnknownError
			}
		}
	}

	func toString() -> String {
		let dict : Dictionary = userInfo
		var message = self.localizedDescription
		let lockey : String = NSError.errorLocationKey()
		if let location = dict[lockey] as? String {
			message = message + "in " + location
		}
		return message
	}
}


